src/main/kotlin/com/sourcegraph/cody/context/RemoteRepoInsight.kt

```diff
@@ -0,0 +1,394 @@
+    val EOF = RemoteRepoTokenType("EOF")
+  }
+}
+
+class RemoteRepoFile(viewProvider: FileViewProvider?) :
+    PsiFileBase(viewProvider!!, RemoteRepoLanguage) {
+  override fun getFileType(): FileType {
+    return RemoteRepoFileType.INSTANCE
+  }
+
+  override fun toString(): String {
+    return "Sourcegraph Remote Repo File"
+  }
+}
+
+private enum class LexerState(val value: Int) {
+  IN_REPO(1),
+  IN_SEPARATOR(2),
+  EOF(3);
+
+  companion object {
+    fun fromInt(value: Int): LexerState? = values().find { it.value == value }
+  }
+}
+
+internal class RemoteRepoListParserDefinition : ParserDefinition {
+  override fun createLexer(project: Project): Lexer {
+    return object : Lexer() {
+      var buffer: CharSequence = ""
+      var startOffset: Int = 0
+      var endOffset: Int = 0
+      var state: LexerState = LexerState.EOF
+      var offset: Int = 0
+
+      override fun start(
+          buffer: CharSequence,
+          startOffset: Int,
+          endOffset: Int,
+          initialState: Int
+      ) {
+        this.buffer = buffer
+        this.startOffset = startOffset
+        this.endOffset = endOffset
+        offset = startOffset
+        state = LexerState.fromInt(initialState) ?: stateAtOffset()
+      }
+
+      override fun getState(): Int {
+        return this.state.value
+      }
+
+      override fun getTokenType(): IElementType? {
+        return when (state) {
+          LexerState.IN_REPO -> RemoteRepoTokenType.REPO
+          LexerState.IN_SEPARATOR -> RemoteRepoTokenType.SEPARATOR
+          LexerState.EOF -> null
+        }
+      }
+
+      override fun getTokenStart(): Int {
+        return this.offset
+      }
+
+      override fun getTokenEnd(): Int {
+        val index =
+            when (tokenType) {
+              RemoteRepoTokenType.REPO ->
+                  buffer.indexOfAny(charArrayOf(' ', '\t', '\r', '\n'), offset)
+              RemoteRepoTokenType.SEPARATOR -> {
+                val subIndex =
+                    buffer.subSequence(offset, buffer.length).indexOfFirst { ch ->
+                      " \t\r\n".indexOf(ch) == -1
+                    }
+                if (subIndex == -1) {
+                  -1
+                } else {
+                  offset + subIndex
+                }
+              }
+              RemoteRepoTokenType.EOF -> return buffer.length
+              else -> throw RuntimeException("unexpected token type $tokenType lexing repo list")
+            }
+        return if (index == -1) {
+          buffer.length
+        } else {
+          index
+        }
+      }
+
+      override fun advance() {
+        this.offset = this.tokenEnd
+        this.state = stateAtOffset()
+      }
+
+      fun stateAtOffset(): LexerState {
+        val ch = nextChar()
+        return when {
+          ch == null -> LexerState.EOF
+          ch.isWhitespace() -> LexerState.IN_SEPARATOR
+          else -> LexerState.IN_REPO
+        }
+      }
+
+      fun nextChar(): Char? {
```

Comment:
Functions like `nextX` usually increment index or move the iterator, so that name could be a bit misleading.